Jeanette/scoot25 job serialization

JeanetteBruno opened the pull request.

CaitieM20 commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,112 @@
+func Deserialize(serializedVal []byte, serializeType string) (*JobDefinition, error) {
+
+	// create the thrift deserializer
+	//deserializer := thrift.NewTDeserializer();
```
nit: comment deserializer := ... no longer necessary?

JeanetteBruno replied:
removed

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+
+import (
+	"testing"
+	//"github.com/leanovate/gopter"
```
nit: delete not needed?

JeanetteBruno replied:
done

CaitieM20 commented on sched/sagaLog.thrift:
```diff
@@ -0,0 +1,16 @@
+
```
Suggestion: Rename JobDef.thrift or Definitions.thrift.  This isn't necessarily specific to the SagaLog.  

The SagaLog just stores opaque bytes doesn't care about the format.  Also this serializer/deserializer will be used to place jobs on the Work Queue too.

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,112 @@
+		return nil, err
+	}
+
+	// make a sched JobDefintion from the thrift JobDefinition for the saga log
```
This seems like a great thing to hoist into a function...

In fact, it might be in scootapi/server/serve.go (in thriftJobToScoot)?

CaitieM20 replied:
The one in serve.go is different its our API format to Scheduler object.  (They are super similar) but we don't want to couple our API object to our internal implementation object.  I would recommend leaving as is

JeanetteBruno replied:
I pulled it into a private method (for readability, not reusability)

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,112 @@
+
+}
+
+func getDeserializer(serializeType string) (*thrift.TDeserializer, error) {
```
You could just have vars for jsonDeserializer and jsonSerializer?

Go makes it super easy to make an interface and then just add a function to an otherwise-empty struct?

JeanetteBruno replied:
done

ryancouto commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,112 @@
+
+const (
+	JsonSerialize string = "json"
+	BinarySerialize string = "binary"
```
nit: running "go fmt" should create whitespace here to line up the two const declarations

JeanetteBruno replied:
done - thanks!

dbentley commented on sched/sagaLog.thrift:
```diff
@@ -0,0 +1,16 @@
+
+struct Command {
```
I'd comment why this is different from scoot.thrift (which we discussed in the chat room, so I think we're all on the same page)

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+	}
+}
+
+func makeSampleJobDefintion() JobDefinition {
```
We already have some helper Job Generators in scoot/sched/generators.go.  
Suggest replacing method with GenJob().JobDefinition.  

Or refactor GenJob() function down into two functions GenJob() and GenJobDef()

CaitieM20 replied:
Oh I see here you want a fixed jobDef not a random one cool!

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+// generate a random string for the type
+func genString(genParams *gopter.GenParameters) string {
```
We have this method in a common directory scoot/tests/testhelpers/generators.go  GenRandomAlphaNumericString() if you want to reuse that!

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+	}
+}
+
+func genJobDefFromParams(genParams *gopter.GenParameters) JobDefinition {
```
Can you move this much better random JobDef function into scoot/sched/generators.go 

That would help make that code better overall!

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+
+}
+
+func genJobDef() gopter.Gen {
```
suggest moving this into scoot/sched/generators.go

CaitieM20 commented on sched/sagaLog.thrift:
```diff
@@ -0,0 +1,16 @@
+struct TaskDefinition {
+  1: Command command,
+}
+
```
Can you add struct Job which mimics Job struct in definitions.go?  We'll need the Id to be stored in the WorkQueue & the SagaLog when we store this information.

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+
+func (this *JobDefinition) Equal(actualJobDef JobDefinition) (bool, string) {
```
Would recommend adding this function to definitions.go for readability.  We follow the convention that all functions on a struct are defined in the same file as the struct definition

dbentley replied:
Oooh, it might be nice, actually, to keep this in _test.go so that it's only included in tests and not in binaries......

CaitieM20 replied:
Its in the same package though, so right now unless we do the test case package aliasing its available to the whole package ... 

I think Equal is a generically useful function.  No reason to isolate to test only.

dbentley replied:
Files named _test.go are included only in the compile for tests. It's like a build directive. So putting it in _test.go (even if in the same package, which, yes, my bad, I've been doing it the wrong way) will include it for tests but not in the binary.

It's fine if we want it in general; I'm -0 on it just because it's creeping featurism and it's not super clear why you'd need it, and definitions of equality often end up confusing later down the road.

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+		return false, fmt.Sprintf("task Definitions maps two different lengths: expected %d, got %d", len(expectedMap), len(actualMap))
+	}
+
+	for taskName, _ := range expectedMap {
```
nit: maybe also create a Equal method on TaskDefinition and add it to definitions.go.  This will be useful in other test cases as well!

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+	return true, ""
+}
+
+func stringSliceCompare(expectedSlice, actualSlice []string) (bool, string) {
```
golang reflect.DeepEqual() might do what you want here without having to write a method to do it.  

http://stackoverflow.com/questions/15311969/checking-the-equality-of-two-slices

JeanetteBruno replied:
good thought - unfortunately DeepEqual() returns false negative when slices and maps are empty

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+
+	// validate comparison logic
+	if ok, msg := origJobDef.Equal(origJobDef); !ok {
+		panic(fmt.Sprintf("error: comparison logic doesn't work comparing an object against itself %s\n", msg))
```
recommend failing test cases with t.Error("error message") or t.Errorf("formated error message")
https://golang.org/pkg/testing/#B.Errorf

This will mark the test case as failed, and allow other test cases to run.  I believe panic will stop all the tests from executing if it fails.

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+func Test_NilJobDefinition(t *testing.T) {
+	if ok, err := Serialize(nil, serializeType); err == nil {
+		panic("error: should have gotten an error from serializing a nil object\n")
+		fmt.Printf(string(ok))
```
nit: would remove the print statement, typically if tests pass there isn't output

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+	return true, ""
+}
+
+func stringMapEqual(expectedMap, actualMap map[string]string) (bool, string) {
```
golang reflect.DeepEqual() might do what you want here without having to write a method to do it.  

http://stackoverflow.com/questions/15311969/checking-the-equality-of-two-slices

CaitieM20 commented on sched/jobDefSerializer_test.go:
```diff
@@ -0,0 +1,249 @@
+	fmt.Printf("\n")
+}
+
+func Test_RandomSerializerDeserializer(t *testing.T) {
```
This is great!

CaitieM20 commented:
Looks good some style comments.  Also you can click the Details link to see why the TravisCI build fails.  

Looks like go vet is failing.  You can run this locally by running make vet from source (excludes vendor/dependency directories)

JeanetteBruno commented:
I did not move the gopter job generation to generators.go because this generator is only for testing.  Note: should GenJobId() gopter.Gen be removed from generators.go? (Intellij is telling me that is isn't used anywhere and the gopter reference makes me think it is only for testing)

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+type JobSerializer struct {
+	writer thrift.TSerializer
```
I'd leave these as pointers

JeanetteBruno replied:
done

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,113 @@
+	schedJobDef.JobType = sagaLogJobDef.JobType
+	schedJobDef.Tasks = make(map[string]TaskDefinition)
+	for taskName, sagaLogTaskDefTask := range sagaLogJobDef.Tasks {
+		argvs := (*(*sagaLogTaskDefTask).Command).Argv
```
Do you need all these *'s? Go lets you dereference pointers automatically with .

(That is, C's . and -> both collapse to . in golang)

JeanetteBruno replied:
I removed them

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// NOTE: leaving this method in this file so it lives close to the Deserializer
+func (job *Job) SerializeJobDef(serializer JobSerializer) ([]byte, error) {
```
Why put this on *Job, as opposed to just having it be a function (not bound to a type) that takes a job and a serializer?

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,113 @@
+package sched
```
Sorry for missing this initially, I think we're trying to prefer foo_bar.go to fooBar.go

JeanetteBruno replied:
ok - making the change

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,113 @@
+
+	writer := thrift.NewTSerializer()
+	reader := thrift.NewTDeserializer()
+	jobSerializer := JobSerializer{*writer, *reader}
```
Dereferencing pointers in Go is a little odd; generally if something gives you a pointer you end up passing it down as a pointer, so maybe JobSerializer could store pointers instead of structs.

JeanetteBruno replied:
fixed

dbentley commented on sched/jobDefSerializer.go:
```diff
@@ -0,0 +1,113 @@
+	for taskName, sagaLogTaskDefTask := range sagaLogJobDef.Tasks {
+		argvs := (*(*sagaLogTaskDefTask).Command).Argv
+		envVars := (*(*sagaLogTaskDefTask).Command).EnvVars
+		timeout, _ := time.ParseDuration(fmt.Sprintf("%dns", (*(*sagaLogTaskDefTask).Command).Timeout))
```
Why sprintf just to parse? This should be just a multiplication (from ms to ns or vice versa?)

JeanetteBruno replied:
this was the only way I could get a Duration object - needed to use time.ParseDuration()

CaitieM20 commented:
LGTM!

dbentley commented:
LGTM

dbentley commented on sched/job_def_serializer.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+type JobSerializer struct {
+	*thrift.TSerializer
```
Give these field names to make it more explicit.

JeanetteBruno commented:
fixed tests to be able to use DeepEqual() and removed Equal() methods in definitions.go

Updated the serialization/deserialization to create new serializers with each request - thrift serialization appears to be state-full giving corrupted results when reused.

Eliminated the need to parse the duration from a string.

dbentley commented on sched/job_def_serializer.go:
```diff
@@ -0,0 +1,136 @@
+
+func (s binarySerializer) Serialize(sourceStruct thrift.TStruct) (b []byte, err error) {
+	d := thrift.NewTSerializer()
+	serializedValue, err := d.Write(sourceStruct)
```
FWIW, you can just:

return d.Write(sourceStruct)

(Also, I might just rename sourceStruct to in)

dbentley commented:
LGTM
